isis/src/mro/apps/hical/hical.cpp

```diff
@@ -0,0 +1,530 @@
+//
+/////////////////////////////////////////////////////////////////////////
+//  Call the processing function
+      /**
+       * @brief Apply calibration to each HiRISE image line
+       *
+       * This function applies the calbration equation to each input image line.  It
+       * gets matrices and constants from the \b calVars container that is established
+       * in the main with some user input via the configuration (CONF) parameter.
+       *
+       * @param in  Input raw image line buffer
+       * @param out Output calibrated image line buffer
+       */
+      auto calibrationFunc = [calVars](Buffer &in, Buffer &out)->void {
+        const HiVector &ZBF    = calVars->get("ZeroBufferFit");
+        const HiVector &ZRev   = calVars->get("ZeroReverse");
+        const HiVector &ZD     = calVars->get("ZeroDark");
+        const HiVector &ZDR    = calVars->get("ZeroDarkRate");
+        const HiVector &GLD    = calVars->get("GainLineDrift");
+        const HiVector &GCN    = calVars->get("GainChannelNormalize");
+        const double GNL       = calVars->get("GainNonLinearity")[0];
+        const HiVector &GFF    = calVars->get("GainFlatField");
+        const HiVector &GT     = calVars->get("GainTemperature");
+        const double GUC       = calVars->get("GainUnitConversion")[0];
+
+        //  Set current line (index)
+        int line(in.Line()-1);
+        if (calVars->exists("LastGoodLine")) {
+          int lastline = ((int) (calVars->get("LastGoodLine"))[0]) - 1;
+          if ( line > lastline ) { line = lastline; }
+        }
+
+
+        //  Apply correction to point of non-linearity accumulating average
+        vector<double> data;
+        for (int i = 0 ; i < in.size() ; i++) {
+          if (IsSpecial(in[i])) {
+            out[i] = in[i];
+          }
+          else {
+            double hdn;
+            hdn = (in[i] - ZBF[line] - ZRev[i] - ZD[i] - ZDR[i]); // Drift, Reverse, Dark, DarkRate
+            hdn = hdn / GLD[line];  // GainLineDrift
+            data.push_back(hdn);   // Accumulate average for non-linearity
+            out[i] = hdn;
+          }
+        }
+
+        // Second loop require to apply non-linearity gain correction from average
+        if ( data.size() > 0 ) {  // No valid pixels means just that - done
+          // See HiCalUtil.h for the function that returns this stat.
+          double NLGain = 1.0 - (GNL * GainLineStat(data));
+          for (int i = 0 ; i < out.size() ; i++) {
+            if (!IsSpecial(out[i])) {
+              double hdn = out[i];
+              hdn = hdn * GCN[i] * NLGain * GFF[i] * GT[i];  // Gain, Non-linearity
+                                                          // gain,  FlatField,
+                                                          // TempGain
+              out[i] = hdn / GUC;                         // I/F or DN or DN/US
+            }
+          }
+        }
+        return;
+      };
```

Comment:
This used to be a function that used a global matrix list so I moved it into a lambda with capture.